src/sync/channel.zig

```diff
@@ -1,170 +1,455 @@
-            defer buffer_lock.unlock();
+        };
 
-            var buffer: *std.ArrayList(T) = buffer_lock.mut();
-            try buffer.append(value);
+        const Buffer = struct {
+            next: Atomic(?*Buffer),
+            slots: [BLOCK_CAP]Slot,
 
-            self.has_value.signal();
-        }
-
-        pub fn sendBatch(self: *Self, value: std.ArrayList(T)) error{ OutOfMemory, ChannelClosed }!void {
-            if (self.closed.load(.monotonic)) {
-                return error.ChannelClosed;
+            fn create(allocator: Allocator) !*Buffer {
+                const new = try allocator.create(Buffer);
+                @memset(&new.slots, Slot.uninit);
+                new.next = Atomic(?*Buffer).init(null);
+                return new;
             }
-            var buffer_lock = self.buffer.lock();
-            defer buffer_lock.unlock();
 
-            var buffer: *std.ArrayList(T) = buffer_lock.mut();
-            try buffer.appendSlice(value.items);
+            fn destroy(block: *Buffer, start: usize, allocator: Allocator) void {
+                for (start..BLOCK_CAP - 1) |i| {
+                    const slot = &block.slots[i];
 
-            self.has_value.signal();
-        }
+                    if (slot.state.load(.acquire) & READ_FROM == 0 and
+                        slot.state.fetchOr(DESTROYED, .acq_rel) & READ_FROM == 0)
+                    {
+                        return;
+                    }
+                }
 
-        pub fn len(self: *Self) error{ChannelClosed}!u64 {
-            if (self.closed.load(.monotonic)) {
-                return error.ChannelClosed;
+                allocator.destroy(block);
             }
-            var buffer = self.buffer.lock();
-            defer buffer.unlock();
 
-            return buffer.get().items.len;
+            fn deinit(block: *Buffer, allocator: Allocator) void {
+                if (block.next.load(.monotonic)) |n| {
+                    n.deinit(allocator);
+                    allocator.destroy(n);
+                }
+            }
+        };
+
+        const Slot = struct {
+            value: T,
+            state: Atomic(usize),
+
+            const uninit: Slot = .{
+                .value = undefined,
+                .state = Atomic(usize).init(0),
+            };
+        };
+
+        pub fn init(allocator: Allocator, initial_capacity: usize) !Self {
+            _ = initial_capacity; // TODO: do something with this
+            const first_block = try Buffer.create(allocator);
+            const first_position: Position = .{
+                .index = Atomic(usize).init(0),
+                .block = Atomic(?*Buffer).init(first_block),
+            };
+            return .{
+                .head = first_position,
+                .tail = first_position,
+                .allocator = allocator,
+            };
+        }
+
+        pub fn create(allocator: Allocator, initial_capacity: usize) !*Self {
```

Comment:
great pattern - `create` vs `init`

Reply:
Agreed. `create` should allocate and return a pointer to `@This()`, `init` should return `@This()` as a value.